autoload/ale/handlers/haskell.vim

```diff
@@ -52,6 +48,7 @@ function! ale#handlers#haskell#HandleGHCFormat(buffer, lines) abort
         endif
 
         call add(l:output, {
+        \   'filename': ale#path#Simplify(getcwd() . '/' . l:match[1]),
```

Comment:
We can't rely on the value of `getcwd()` here. This function could be run from any buffer, and the current directory might not be the directory the file is in.

Are all of the paths here relative to the file being checked? I don't know what the filenames look like. If they are, then you should use `expand('#' . a:buffer . ':p:h')` to get the directory that the file being checked is in. I recommend saving that value in a variable above.

Play around with something like this and test it.

Reply:
> Are all of the paths here relative to the file being checked?

No, they are relative to the cwd of the `ghc` (or `ghc-mod` etc.) process. Is there an easy way to find this directory?

Reply:
In that case, you can do what was done for some other linters. You can use the commands for adding `cd ... &&` to the front of the commands so the directory `ghc` will run from will be predictable and then resolve the paths from that same directory in the callback for building the loclist items.

Here is an example from the tslint.vim file.

```vim
" The command always switches to the directory for the buffer explicitly.
    return ale#path#BufferCdString(a:buffer) . ...

... 

" The directory is computed once at the top of the handler.
function! ale_linters#typescript#tslint#Handle(buffer, lines) abort
    let l:dir = expand('#' . a:buffer . ':p:h')

...

" The absolute path is resolved using the working directory.
        \   'filename': ale#path#GetAbsPath(l:dir, l:error.name),
```

Hopefully that should help.

Reply:
Hmm, I think it can be very surprising to the user if `ale` silently used a different working directory for running these commands, because it can mess up things like search paths. It will be an even bigger problem for `hdevtools`, one of the ghc-based linters, because it uses a unix domain socket in the current directory to communicate with the server.

Would it be acceptable to extend the `job_info` dictionary to include the cwd of the job, and pass it over to the linter callback upon completion?

Reply:
No, it should be fine. Other linters already do the same thing, and you can see the `cd` command in `ALEInfo`. Try it, and see what happens.

Reply:
I just tried it and it did not work. The code can be found here: https://github.com/takano-akio/ale/commit/96a6e9720078ba577c1197de42d904d8f46ec4c5

Let me explain the problem: imagine a directory hierarchy like this:

```
src/
  main.hs # defines the module Main, which imports Lib.Foo
  Lib/
    Foo.hs # defines the module Lib.Foo, which imports Lib.Bar
    Bar.hs # defines the module Lib.Bar
```

I open vim with cwd = `src/`, and do `:e Lib/Foo.hs`. Then `ale` invokes `ghc` on this file. Without a `cd` this works fine. With a `cd` this doesn't work, because `ghc` fails to find the module `Lib.Bar`. According to the Haskell convention, the module `Lib.Bar` must be found in a file called `Lib/Bar.hs` relative to some source directory, and by default the only source directory is `.`. If `ale` issues a `cd` into the `Lib` directory before invoking `ghc`, `ghc` will look for `src/Lib/Lib/Bar.hs`, which doesn't exist.

Reply:
Okay cool. Change the directory to the project directory then.

Reply:
There is no universal way to find the project directory. There sometimes is, depending on how the project is set up, but not always. Hence my suggestion of recording the cwd.

Reply:
Use whatever that directory is, but you can't use `getcwd()`. The function for handling errors can be run from other unrelated buffers, and users can set their working directory to any directory on their filesystem.

If there's no one way to find the project directory, then we should try and detect it and make it configurable, so we can find it most of the time and let users explicitly set it when we can't figure out out automatically.